attached_assets/Pasted--settings-extra-key-combining-caron-false-longpress-int_1767755018962.txt
{"settings":{"extra_key_combining_caron":false,"longpress_interval":12,"extra_key_combining_shaddah":false,"margin_bottom_portrait_unfolded":7,"extra_key_page_up":true,"key_horizontal_margin":2,"extra_key_cut":true,"circle_sensitivity":"2","extra_key_capslock":true,"extra_key_combining_fatha":false,"layouts":"[{\"kind\":\"system\"}]","extra_key_†":false,"key_activated_opacity":100,"extra_key_combining_grave":false,"extra_key_accent_trema":false,"extra_key_nbsp":false,"extra_key_accent_breve":false,"number_entry_layout":"pin","keyboard_opacity":100,"border_config":false,"extra_key_combining_sukun":false,"extra_key_delete_word":false,"extra_key_combining_hamza_above":false,"character_size":1.15,"extra_key_combining_kasratan":false,"extra_key_accent_double_grave":false,"extra_key_accent_slash":false,"extra_key_combining_double_aigu":false,"custom_border_line_width":0,"extra_key_combining_horn":false,"version":3,"swipe_dist":"15","extra_key_switch_greekmath":false,"extra_key_combining_payerok":false,"extra_key_switch_clipboard":true,"extra_key_accent_tilde":false,"keyboard_height_landscape":50,"keyboard_height_unfolded":35,"extra_key_accent_dot_above":false,"extra_key_combining_inverted_breve":false,"extra_key_combining_ogonek":false,"lock_double_tap":true,"extra_key_shareText":true,"extra_key_combining_slavonic_psili":false,"extra_key_combining_vzmet":false,"number_row":"no_number_row","extra_key_combining_titlo":false,"keyboard_height":26,"extra_key_compose":true,"extra_key_undo":true,"extra_key_scroll_lock":true,"extra_key_accent_cedille":false,"extra_key_ß":false,"extra_key_accent_grave":false,"horizontal_margin_portrait":3,"extra_key_esc":true,"extra_key_combining_tilde":false,"margin_bottom_portrait":7,"extra_key_page_down":true,"theme":"jungle","extra_key_home":false,"extra_key_accent_hook_above":false,"extra_key_accent_horn":false,"extra_key_combining_cedille":false,"extra_key_meta":false,"slider_sensitivity":"30","extra_key_combining_breve":false,"vibrate_duration":41,"extra_key_combining_slash":false,"numpad_layout":"high_first","extra_key_combining_trema":false,"extra_key_combining_alef_below":false,"switch_input_immediate":false,"extra_key_redo":true,"longpress_timeout":600,"extra_key_accent_circonflexe":false,"extra_key_combining_dot_below":false,"show_numpad":"1","margin_bottom_landscape":3,"key_vertical_margin":1.5,"extra_key_€":false,"extra_key_combining_ring":false,"extra_key_§":false,"extra_key_change_method":true,"need_migration":false,"extra_key_ª":false,"extra_key_combining_circonflexe":false,"extra_key_£":false,"keyboard_height_landscape_unfolded":50,"extra_key_zwj":false,"extra_key_accent_dot_below":false,"autocapitalisation":true,"extra_key_combining_aigu":false,"extra_key_superscript":false,"extra_key_combining_palatalization":false,"extra_key_combining_fathatan":false,"extra_key_combining_bar":false,"extra_key_combining_vertical_tilde":false,"keyrepeat_enabled":true,"extra_key_f12_placeholder":true,"horizontal_margin_landscape":28,"extra_key_f11_placeholder":true,"extra_key_combining_kasra":false,"extra_key_accent_ogonek":false,"extra_key_alt":true,"extra_key_combining_alef_above":false,"extra_key_combining_dammah":false,"extra_key_pasteAsPlainText":true,"extra_key_combining_kavyka":false,"extra_key_tab":true,"extra_key_combining_hamza_below":false,"extra_key_zwnj":false,"extra_key_º":false,"margin_bottom_landscape_unfolded":3,"extra_key_nnbsp":false,"extra_key_combining_hook_above":false,"extra_key_accent_caron":false,"extra_key_accent_double_aigu":false,"extra_key_combining_dammatan":false,"vibrate_custom":true,"extra_key_combining_arabic_inverted_v":false,"extra_key_combining_arrow_right":false,"label_brightness":100,"extra_key_accent_aigu":false,"horizontal_margin_landscape_unfolded":28,"extra_key_selectAll":true,"extra_key_combining_dot_above":false,"extra_key_combining_arabic_v":false,"extra_key_accent_bar":false,"extra_key_menu":true,"key_opacity":100,"clipboard_history_enabled":true,"custom_border_radius":0,"extra_key_subscript":false,"extra_key_accent_macron":false,"extra_key_combining_macron":false,"extra_key_paste":true,"extra_key_forward_delete_word":false,"extra_key_accent_ring":false,"extra_key_copy":true,"extra_key_combining_pokrytie":false,"extra_key_end":false,"horizontal_margin_portrait_unfolded":3,"extra_key_combining_slavonic_dasia":false,"extra_key_voice_typing":true},"clipboard":[{"content":"ketones due to steric reasons. The hydrogensulphite addition compound is water soluble and can be converted back to the original carbonyl compound by treating it with dilute mineral acid or alkali. Therefore, these are useful for separation and purification of aldehydes. \n(c) Addition of Grignard reagents: (refer Unit 7, Class XII). (d) Addition of alcohols: Aldehydes react with one equivalent of \nmonohydric alcohol in the presence of dry hydrogen chloride to yield alkoxyalcohol intermediate, known as hemiacetals, which further react with one more molecule","timestamp":"2026-01-06 03:20:25","description":"","version":""},{"content":"compounds","timestamp":"2026-01-06 03:20:25","description":"","version":""},{"content":"water","timestamp":"2026-01-06 03:20:25","description":"","version":""},{"content":"Sterically, the presence of two relatively","timestamp":"2026-01-06 03:20:25","description":"","version":""},{"content":"chemical reactions. \n1. Nucleophilic addition reactions \nContrary to electrophilic addition reactions observed in alkenes, the aldehydes and ketones undergo nucleophilic addition reactions. (i) Mechanism of nucleophilic addition reactions \nA nucleophile attacks the electrophilic carbon atom of the polar carbonyl group from a direction approximately perpendicular to the plane of sp2 hybridised orbitals of carbonyl carbon (Fig. 8.2). The hybridisation of carbon changes from sp2 to sp3 in this process, and a tetrahedral alkoxide intermediate is produced. This intermediate captures a proton from the \nreaction medium to give the electrically neutral product. The net result is addition of Nu– and H \n+ \nacross the carbon oxygen double bond as shown in Fig. 8.2. \nIntext Question 8.3 Arrange the following compounds in increasing order of \ntheir","timestamp":"2026-01-06 03:20:25","description":"","version":""},{"content":"boiling points. CH3CHO, CH3CH2OH, CH3OCH3, CH3CH2CH3","timestamp":"2026-01-06 03:20:25","description":"","version":""},{"content":"range of 72 to 74. Since only butan-1-ol molecules are associated due to extensive intermolecular hydrogen bonding","timestamp":"2026-01-06 03:20:25","description":"","version":""},{"content":"and ketones decreases rapidly on increasing the length of alkyl chain. All aldehydes and ketones are fairly soluble in organic solvents like benzene, ether, methanol, chloroform, etc. The lower aldehydes have sharp pungent odours. As the size of the molecule increases, the odour becomes less pungent and more fragrant. In fact, many naturally occurring aldehydes and ketones are used in the blending of perfumes and flavouring agents. ","timestamp":"2026-01-06 03:20:25","description":"","version":""},{"content":"Whenever something","timestamp":"2026-01-06 10:29:55","description":"","version":""}]}

srcs/juloo.keyboard2/Config.java
package juloo.keyboard2;

import android.content.Context;
import android.content.SharedPreferences;
import android.content.res.Configuration;
import android.content.res.Resources;
import android.util.DisplayMetrics;
import android.util.TypedValue;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import juloo.keyboard2.prefs.CustomExtraKeysPreference;
import juloo.keyboard2.prefs.ExtraKeysPreference;
import juloo.keyboard2.prefs.LayoutsPreference;

public final class Config
{
  /**
   * Width of the Android phones is around 300-600dp in portrait, 600-1400dp in landscape,
   * depending on the user's size settings.
   *
   * 600 dp seems a reasonable midpoint to determine whether the current orientation of the device is "wide"
   * (landsacpe, tablet, unfolded foldable etc.) or not, to switch to a different layout.
   */
  public static final int WIDE_DEVICE_THRESHOLD = 600;

  private final SharedPreferences _prefs;

  // From resources
  public final float marginTop;
  public final float keyPadding;

  public final float labelTextSize;
  public final float sublabelTextSize;

  // From preferences
  /** [null] represent the [system] layout. */
  public List<KeyboardData> layouts;
  public boolean show_numpad = false;
  // From the 'numpad_layout' option, also apply to the numeric pane.
  public boolean inverse_numpad = false;
  public boolean add_number_row;
  public boolean number_row_symbols;
  public float swipe_dist_px;
  public float slide_step_px;
  // Let the system handle vibration when false.
  public boolean vibrate_custom;
  // Control the vibration if [vibrate_custom] is true.
  public long vibrate_duration;
  public long longPressTimeout;
  public long longPressInterval;
  public boolean keyrepeat_enabled;
  public float margin_bottom;
  public int keyboardHeightPercent;
  public int screenHeightPixels;
  public float horizontal_margin;
  public float key_vertical_margin;
  public float key_horizontal_margin;
  public int labelBrightness; // 0 - 255
  public int keyboardOpacity; // 0 - 255
  public float customBorderRadius; // 0 - 1
  public float customBorderLineWidth; // dp
  public int keyOpacity; // 0 - 255
  public int keyActivatedOpacity; // 0 - 255
  public boolean double_tap_lock_shift;
  public float characterSize; // Ratio
  public int theme; // Values are R.style.*
  public boolean autocapitalisation;
  public boolean switch_input_immediate;
  public NumberLayout selected_number_layout;
  public boolean borderConfig;
  public int circle_sensitivity;
  public boolean clipboard_history_enabled;
  public int clipboard_history_duration;

  // Dynamically set
  /** Configuration options implied by the connected editor. */
  public EditorConfig editor_config;
  public boolean should_show_candidates_view;
  public boolean shouldOfferVoiceTyping;
  public ExtraKeys extra_keys_subtype;
  public Map<KeyValue, KeyboardData.PreferredPos> extra_keys_param;
  public Map<KeyValue, KeyboardData.PreferredPos> extra_keys_custom;

  public final IKeyEventHandler handler;
  public boolean orientation_landscape = false;
  public boolean foldable_unfolded = false;
  public boolean wide_screen = false;
  /** Index in 'layouts' of the currently used layout. See
      [get_current_layout()] and [set_current_layout()]. */
  int current_layout_narrow;
  int current_layout_wide;

  private Context _context;

  public void setContext(Context ctx) {
    _context = ctx;
  }

  public Context getContext() {
    return _context;
  }

  private Config(SharedPreferences prefs, Resources res, IKeyEventHandler h, Boolean foldableUnfolded)
  {
    _prefs = prefs;
    editor_config = new EditorConfig();
    // static values
    marginTop = res.getDimension(R.dimen.margin_top);
    keyPadding = res.getDimension(R.dimen.key_padding);
    labelTextSize = 0.33f;
    sublabelTextSize = 0.22f;
    // from prefs
    refresh(res, foldableUnfolded);
    // initialized later
    should_show_candidates_view = false;
    shouldOfferVoiceTyping = false;
    extra_keys_subtype = null;
    handler = h;
  }

  /*
   ** Reload prefs
   */
  public void refresh(Resources res, Boolean foldableUnfolded)
  {
    DisplayMetrics dm = res.getDisplayMetrics();
    orientation_landscape = res.getConfiguration().orientation == Configuration.ORIENTATION_LANDSCAPE;
    foldable_unfolded = foldableUnfolded;

    float characterSizeScale = 1.f;
    String show_numpad_s = _prefs.getString("show_numpad", "never");
    show_numpad = "always".equals(show_numpad_s);
    if (orientation_landscape)
    {
      if ("landscape".equals(show_numpad_s))
        show_numpad = true;
      keyboardHeightPercent = _prefs.getInt(foldable_unfolded ? "keyboard_height_landscape_unfolded" : "keyboard_height_landscape", 50);
      characterSizeScale = 1.25f;
    }
    else
    {
      keyboardHeightPercent = _prefs.getInt(foldable_unfolded ? "keyboard_height_unfolded" : "keyboard_height", 35);
    }
    layouts = LayoutsPreference.load_from_preferences(res, _prefs);
    inverse_numpad = _prefs.getString("numpad_layout", "default").equals("low_first");
    String number_row = _prefs.getString("number_row", "no_number_row");
    add_number_row = !number_row.equals("no_number_row");
    number_row_symbols = number_row.equals("symbols");
    // The baseline for the swipe distance correspond to approximately the
    // width of a key in portrait mode, as most layouts have 10 columns.
    // Multipled by the DPI ratio because most swipes are made in the diagonals.
    // The option value uses an unnamed scale where the baseline is around 25.
    float dpi_ratio = Math.max(dm.xdpi, dm.ydpi) / Math.min(dm.xdpi, dm.ydpi);
    float swipe_scaling = Math.min(dm.widthPixels, dm.heightPixels) / 10.f * dpi_ratio;
    float swipe_dist_value = Float.valueOf(_prefs.getString("swipe_dist", "15"));
    swipe_dist_px = swipe_dist_value / 25.f * swipe_scaling;
    float slider_sensitivity = Float.valueOf(_prefs.getString("slider_sensitivity", "30")) / 100.f;
    slide_step_px = slider_sensitivity * swipe_scaling;
    vibrate_custom = _prefs.getBoolean("vibrate_custom", true);
    vibrate_duration = _prefs.getInt("vibrate_duration", 41);
    longPressTimeout = _prefs.getInt("longpress_timeout", 600);
    longPressInterval = _prefs.getInt("longpress_interval", 12);
    keyrepeat_enabled = _prefs.getBoolean("keyrepeat_enabled", true);
    margin_bottom = get_dip_pref_oriented(dm, "margin_bottom", 7, 3);
    key_vertical_margin = get_dip_pref(dm, "key_vertical_margin", 1.5f) / 100;
    key_horizontal_margin = get_dip_pref(dm, "key_horizontal_margin", 2) / 100;
    // Label brightness is used as the alpha channel
    labelBrightness = _prefs.getInt("label_brightness", 100) * 255 / 100;
    // Keyboard opacity
    keyboardOpacity = _prefs.getInt("keyboard_opacity", 100) * 255 / 100;
    keyOpacity = _prefs.getInt("key_opacity", 100) * 255 / 100;
    keyActivatedOpacity = _prefs.getInt("key_activated_opacity", 100) * 255 / 100;
    // keyboard border settings
    borderConfig = _prefs.getBoolean("border_config", false);
    customBorderRadius = _prefs.getInt("custom_border_radius", 0) / 100.f;
    customBorderLineWidth = get_dip_pref(dm, "custom_border_line_width", 0);
    screenHeightPixels = dm.heightPixels;
    horizontal_margin =
      get_dip_pref_oriented(dm, "horizontal_margin", 3, 28);
    double_tap_lock_shift = _prefs.getBoolean("lock_double_tap", true);
    characterSize =
      _prefs.getFloat("character_size", 1.15f)
      * characterSizeScale;
    theme = getThemeId(res, _prefs.getString("theme", ""));
    autocapitalisation = _prefs.getBoolean("autocapitalisation", true);
    switch_input_immediate = _prefs.getBoolean("switch_input_immediate", false);
    extra_keys_param = ExtraKeysPreference.get_extra_keys(_prefs);
    extra_keys_custom = CustomExtraKeysPreference.get(_prefs);
    selected_number_layout = NumberLayout.of_string(_prefs.getString("number_entry_layout", "pin"));
    current_layout_narrow = _prefs.getInt("current_layout_portrait", 0);
    current_layout_wide = _prefs.getInt("current_layout_landscape", 0);
    circle_sensitivity = Integer.valueOf(_prefs.getString("circle_sensitivity", "2"));
    clipboard_history_enabled = _prefs.getBoolean("clipboard_history_enabled", false);
    clipboard_history_duration = Integer.parseInt(_prefs.getString("clipboard_history_duration", "5"));

    float screen_width_dp = dm.widthPixels / dm.density;
    wide_screen = screen_width_dp >= WIDE_DEVICE_THRESHOLD;
  }

  public int get_current_layout()
  {
    return (wide_screen)
            ? current_layout_wide : current_layout_narrow;
  }

  public void set_current_layout(int l)
  {
    if (wide_screen)
      current_layout_wide = l;
    else
      current_layout_narrow = l;

    SharedPreferences.Editor e = _prefs.edit();
    e.putInt("current_layout_portrait", current_layout_narrow);
    e.putInt("current_layout_landscape", current_layout_wide);
    e.apply();
  }

  public void set_clipboard_history_enabled(boolean e)
  {
    clipboard_history_enabled = e;
    _prefs.edit().putBoolean("clipboard_history_enabled", e).apply();
  }

  private float get_dip_pref(DisplayMetrics dm, String pref_name, float def)
  {
    float value;
    try { value = _prefs.getInt(pref_name, -1); }
    catch (Exception e) { value = _prefs.getFloat(pref_name, -1f); }
    if (value < 0f)
      value = def;
    return (TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, value, dm));
  }

  /** [get_dip_pref] depending on orientation. */
  float get_dip_pref_oriented(DisplayMetrics dm, String pref_base_name, float def_port, float def_land)
  {
    final String suffix;
    if (foldable_unfolded) {
      suffix = orientation_landscape ? "_landscape_unfolded" : "_portrait_unfolded";
    } else {
      suffix = orientation_landscape ? "_landscape" : "_portrait";
    }

    float def = orientation_landscape ? def_land : def_port;
    return get_dip_pref(dm, pref_base_name + suffix, def);
  }

  private int getThemeId(Resources res, String theme_name)
  {
    int night_mode = res.getConfiguration().uiMode & Configuration.UI_MODE_NIGHT_MASK;
    switch (theme_name)
    {
      case "light": return R.style.Light;
      case "black": return R.style.Black;
      case "altblack": return R.style.AltBlack;
      case "dark": return R.style.Dark;
      case "white": return R.style.White;
      case "epaper": return R.style.ePaper;
      case "desert": return R.style.Desert;
      case "jungle": return R.style.Jungle;
      case "monetlight": return R.style.MonetLight;
      case "monetdark": return R.style.MonetDark;
      case "monet":
        if ((night_mode & Configuration.UI_MODE_NIGHT_NO) != 0)
          return R.style.MonetLight;
        return R.style.MonetDark;
      case "rosepine": return R.style.RosePine;
      case "everforestlight": return R.style.EverforestLight;
      case "cobalt": return R.style.Cobalt;
      case "pine": return R.style.Pine;
      case "epaperblack": return R.style.ePaperBlack;
      default:
      case "system":
        if ((night_mode & Configuration.UI_MODE_NIGHT_NO) != 0)
          return R.style.Light;
        return R.style.Dark;
    }
  }

  private static Config _globalConfig = null;

  public static void initGlobalConfig(SharedPreferences prefs, Resources res,
      IKeyEventHandler handler, Boolean foldableUnfolded)
  {
    migrate(prefs);
    _globalConfig = new Config(prefs, res, handler, foldableUnfolded);
    LayoutModifier.init(_globalConfig, res);
  }

  public static Config globalConfig()
  {
    return _globalConfig;
  }

  public static SharedPreferences globalPrefs()
  {
    return _globalConfig._prefs;
  }

  public static interface IKeyEventHandler
  {
    public void key_down(KeyValue value, boolean is_swipe);
    public void key_up(KeyValue value, Pointers.Modifiers mods);
    public void mods_changed(Pointers.Modifiers mods);
    public void suggestion_entered(String text);
  }

  /** Config migrations. */

  private static int CONFIG_VERSION = 3;

  public static void migrate(SharedPreferences prefs)
  {
    int saved_version = prefs.getInt("version", 0);
    Logs.debug_config_migration(saved_version, CONFIG_VERSION);
    if (saved_version == CONFIG_VERSION)
      return;
    SharedPreferences.Editor e = prefs.edit();
    e.putInt("version", CONFIG_VERSION);
    // Migrations might run on an empty [prefs] for new installs, in this case
    // they set the default values of complex options.
    switch (saved_version)
    {
      case 0:
        // Primary, secondary and custom layout options are merged into the new
        // Layouts option. This also sets the default value.
        List<LayoutsPreference.Layout> l = new ArrayList<LayoutsPreference.Layout>();
        l.add(migrate_layout(prefs.getString("layout", "system")));
        String snd_layout = prefs.getString("second_layout", "none");
        if (snd_layout != null && !snd_layout.equals("none"))
          l.add(migrate_layout(snd_layout));
        String custom_layout = prefs.getString("custom_layout", "");
        if (custom_layout != null && !custom_layout.equals(""))
          l.add(LayoutsPreference.CustomLayout.parse(custom_layout));
        LayoutsPreference.save_to_preferences(e, l);
        // Fallthrough
      case 1:
        boolean add_number_row = prefs.getBoolean("number_row", false);
        e.putString("number_row", add_number_row ? "no_symbols" : "no_number_row");
        // Fallthrough
      case 2:
        if (!prefs.contains("number_entry_layout")) {
          e.putString("number_entry_layout", prefs.getBoolean("pin_entry_enabled", true) ? "pin" : "number");
        }
        // Fallthrough
      case 3:
      default: break;
    }
    e.apply();
  }

  private static LayoutsPreference.Layout migrate_layout(String name)
  {
    if (name == null || name.equals("system"))
      return new LayoutsPreference.SystemLayout();
    return new LayoutsPreference.NamedLayout(name);
  }
}
